GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,274 @@
+        {
+            if (healthData.GitTrackedItemsCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)healthData.PlaceholderCount / healthData.GitTrackedItemsCount;
+        }
+
+        public decimal GetModifiedPathsPercentage(GVFSEnlistmentHealthData healthData)
+        {
+            if (healthData.GitTrackedItemsCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)healthData.ModifiedPathsCount / healthData.GitTrackedItemsCount;
+        }
+
+        public decimal CalculateHealthMetric(GVFSEnlistmentHealthData healthData)
+        {
+            return this.CalculateHealthMetric(healthData.PlaceholderCount + healthData.ModifiedPathsCount, healthData.GitTrackedItemsCount);
+        }
+
+        /// <summary>
+        /// Take a string representing a file path on system and pull out the upper-most directory from it as a string, or GVFSConstants.GitPathSeparator if it is in the root
+        /// </summary>
+        /// <param name="path">The path to a file to parse for the top level directory containing it</param>
+        /// <returns>A string containing the top level directory from the provided path, or GVFSConstants.GitPathSeparator if the path is for an item in the root</returns>
+        private string ParseTopDirectory(string path)
+        {
+            int whackLocation = path.IndexOf(GVFSConstants.GitPathSeparator);
+            if (whackLocation == -1)
+            {
+                return GVFSConstants.GitPathSeparatorString;
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it doesn't count towards its own hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        /// <param name="parentDirectory">Paths will only be categorized if they are descendants of the parentDirectory</param>
+        /// <param name="isFile">Flag which specifies whether the list of paths are paths to files (true) or directories (false)</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory, bool isFile)

```

Comment:
I realized that I was accounting for the same special case of being just under the top level directory, and that with the changes to counting folders this morning a lot of it could be simplified. Now the distinction between folders and files doesn't exist, and all logic is shared.